singer/utils.py

```diff
@@ -184,45 +188,51 @@ def parse_args(required_config_keys):
     return args
 
 
-def check_config(config, required_keys):
+def check_config(config: dict, required_keys: List[str]) -> None:
     missing_keys = [key for key in required_keys if key not in config]
     if missing_keys:
         raise Exception(f'Config is missing required keys: {missing_keys}')
 
 
-def backoff(exceptions, giveup):
+def backoff(exceptions, giveup) -> Callable:
     """Decorates a function to retry up to 5 times using an exponential backoff
     function.
 
     exceptions is a tuple of exception classes that are retried
     giveup is a function that accepts the exception and returns True to retry
     """
-    return backoff_module.on_exception(
-        backoff_module.expo,
-        exceptions,
-        max_tries=5,
-        giveup=giveup,
-        factor=2)
-
-
-def exception_is_4xx(exception):
+    return cast(
+        Callable,
+        backoff_module.on_exception(
+            backoff_module.expo,
+            exceptions,
+            max_tries=5,
+            giveup=giveup,
+            factor=2
+        )
+    )
+
+
+def exception_is_4xx(exception: Exception) -> bool:
     """Returns True if exception is in the 4xx range."""
     if not hasattr(exception, 'response'):
         return False
 
-    if exception.response is None:
+    response = exception.response  # type: ignore  # Duck-typed requests.Response
+
+    if response is None:
         return False
 
     if not hasattr(exception.response, 'status_code'):
```

Comment:
Use already resolved `response` var from above.

```suggestion
    if not hasattr(response, 'status_code'):
```